File: course content/1_computing_intro/Kernfeld_Eric_FizzBuzz.R
# Example solution to FizzBuzz in R for Eric's HEART class on principles of statistics. 

# Do it for just one number
FizzBuzzOne = function(i){
    output = ""
    if( i %% 3 == 0 ){
        output = paste0(output, "Fizz")
    }
    if( i %% 5 == 0 ){
        output = paste0(output, "Buzz")
    }
    if( nchar(output)==0 ){
        output = paste0(output, i)
    }
    output
}
# Apply it to all numbers
FizzBuzz = function(N) sapply(1:N, FizzBuzzOne)

# Test cases
FizzBuzzOne(3)
FizzBuzzOne(6)
FizzBuzzOne(5)
FizzBuzzOne(10)
FizzBuzzOne(15)
FizzBuzzOne(0)
FizzBuzzOne(-3)
FizzBuzzOne(60)
FizzBuzz(15)
# Test under mis-use or abuse
FizzBuzzOne("axolotl")
FizzBuzz("axolotls are cool")
FizzBuzz(1:15)

# For exercise 4
concatenated_output = paste0(FizzBuzz(1000), collapse = "")
character_frequencies = table(strsplit(concatenated_output, "")[[1]])
character_frequencies = setNames(as.data.frame(character_frequencies), c("Character", "Frequency"))
write.csv(character_frequencies, "character_frequencies.csv") #save it to a file

# Exercise 5
library(ggplot2)
ggplot(character_frequencies) + 
    geom_point(aes(x = Character, y = Frequency)) + 
    ggtitle("Character frequencies in FizzBuzz(1000)")
ggsave("example_plot1.png", width = 5, height = 5)
ggplot(character_frequencies) + 
    geom_bar(aes(x = Character, y = Frequency), stat = "identity") + 
    ggtitle("Character frequencies in FizzBuzz(1000)")
ggsave("example_plot2.png", width = 5, height = 5)
{
    png("example_plot3.png", width = 500, height = 500)
    pie(character_frequencies$Frequency, character_frequencies$Character, main = "Character frequencies in FizzBuzz(1000)")
    dev.off()
}
ggplot(character_frequencies) + 
    geom_bar(aes(x = "", y = Frequency, fill = Character), stat = "identity", position = "stack") + 
    ggtitle("Character frequencies in FizzBuzz(1000)") 
ggsave("example_plot4.png", width = 5, height = 5)

ggplot(character_frequencies) + 
    geom_bar(aes(x = "", y = Frequency, fill = Character), stat = "identity", position = "stack") + 
    ggtitle("Character frequencies in FizzBuzz(1000)", "...with a questionable labeling scheme") + 
    scale_fill_manual(values = rep(c("grey", "black"), length.out = 15))
ggsave("example_plot5.png", width = 5, height = 5)

# Common problem #1: using string literals in place of variable names
ggplot(character_frequencies) + 
  geom_bar(aes(x = "", y = Frequency,   fill = Character), stat = "identity", position = "stack") # right
ggplot(character_frequencies) + 
  geom_bar(aes(x = "", y = "Frequency", fill = "Character"), stat = "identity", position = "stack") # wrong
# Common problem #2: fixing color or fill inside vs outside aes().
# Suppose you want all chart elements to be colored pink.
ggplot(character_frequencies) + 
  geom_bar(aes(x = "", y = Frequency), fill = "pink", stat = "identity", position = "stack") # right
ggplot(character_frequencies) + 
  geom_bar(aes(x = "", y = Frequency, fill = "pink"), stat = "identity", position = "stack") # wrong

File: course content/1_computing_intro/Kernfeld_Eric_FizzBuzz.py
# Do it for just one number
def FizzBuzzOne(i):
    output = ""
    if i % 3 == 0:
        output += "Fizz"
    if i % 5 == 0:
        output += "Buzz"
    if len(output) == 0:
        output += str(i)
    return output

# Apply it to all numbers with a list comprehension
def FizzBuzz(N):
    return [FizzBuzzOne(i) for i in range(1, N+1)]

# Here's another way to do the same thing.
def FizzBuzzLoop(N):
    result = [None] * N
    for i in range(1, N+1):
        result[i-1] = FizzBuzzOne(i)
    return result

# Test cases
print(FizzBuzzOne(3))
print(FizzBuzzOne(6))
print(FizzBuzzOne(5))
print(FizzBuzzOne(10))
print(FizzBuzzOne(15))
print(FizzBuzzOne(0))
print(FizzBuzzOne(-3))
print(FizzBuzzOne(60))
print(FizzBuzz(15))

# For exercise 4
import pandas as pd

concatenated_output = ''.join(FizzBuzz(1000))
character_frequencies = pd.Series(list(concatenated_output)).value_counts().reset_index()
character_frequencies.columns = ['Character', 'Frequency']
character_frequencies.to_csv("character_frequencies.csv", index=False)

# Exercise 5
import matplotlib.pyplot as plt

character_frequencies = character_frequencies.sort_values('Character')
plt.figure()
plt.scatter(character_frequencies['Character'], character_frequencies['Frequency'])
plt.title("Character frequencies in FizzBuzz(1000)")
plt.savefig("example_plot1.png")

plt.figure()
plt.bar(character_frequencies['Character'], character_frequencies['Frequency'])
plt.title("Character frequencies in FizzBuzz(1000)")
plt.savefig("example_plot2.png")

plt.figure()
plt.pie(character_frequencies['Frequency'], labels=character_frequencies['Character'], autopct='%1.1f%%')
plt.title("Character frequencies in FizzBuzz(1000)")
plt.savefig("example_plot3.png")

plt.figure()
plt.bar(height = character_frequencies['Frequency'], x=character_frequencies['Character'])
plt.title("Character frequencies in FizzBuzz(1000)")
plt.savefig("example_plot4.png")
